persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/StargateQueryHandler.java

```diff
@@ -150,10 +150,14 @@ public ResultMessage process(
       String s, ClientState clientState, Map<String, ByteBuffer> map)
       throws RequestValidationException {
     ResultMessage.Prepared prepare = QueryProcessor.instance.prepare(s, clientState, map);
-    Prepared prepared = QueryProcessor.instance.getPrepared(prepare.statementId);
-    boolean idempotent = IdempotencyAnalyzer.isIdempotent(prepared.statement);
-    boolean useKeyspace = prepared.statement instanceof UseStatement;
-    return new PreparedWithInfo(idempotent, useKeyspace, prepare);
+    CQLStatement statement =
+        Optional.ofNullable(QueryProcessor.instance.getPrepared(prepare.statementId))
+            .map(p -> p.statement)
+            .orElseGet(() -> QueryProcessor.getStatement(s, clientState));
```

Comment:
Yes `QueryProcessor.getStatement` re-parse and re-prepares, with out interacting with the cache. In addition we get a statement back. Instead of the MD5Digest.

Cache will get evicted and then re-populated by another thread. Last thread always inserts to cache.